### PR TITLE
[rai] Fix extraction for recent raiplay.it updates

### PR DESCRIPTION
**Pull request by iamleot**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

- Introduce _BASE_URL in RaiBaseIE class so it could be reused as base for the
  several subextractors.
- Remove first test of RaiPlayIE: it is no longer available
- Make RaiPlayIE extension-agnostic (passing possible `.json' URLs is now
  supported too)
- Adjust RaiPlayLiveIE to recent raiplay.it updates.  Passing it as
  `url_transparent' is no longer supported (there is no longer an accessible
  ContentItem)
- Adjust RaiPlayPlaylistIE to recent raiplay.it updates and instruct it about
  ContentSet-s.
- Update a RaiIE test and remove two tests that are no longer availables

This fix issue #22923, #22906 and supersedes #23006 and #23040.

### CODE REVIEW

**Comment by iamleot** [comment]
JFTR, it seems that the all failed checks are due the Travis CI build that was cancelled (so the code touched by the PR was actually not tested).

File: youtube_dl/extractor/rai.py [hunk 1]
```diff
@@ -240,20 +222,32 @@ class RaiPlayLiveIE(RaiBaseIE):
...
+        info = {
             'id': video_id,
             'display_id': display_id,
+            'title': self._live_title(title) if relinker_info.get(
+                'is_live') else title,
+            'alt_title': media.get('subtitle'),
+            'description': media.get('description'),
+            'uploader': strip_or_none(media.get('channel')),
+            'creator': strip_or_none(media.get('editor')),
+            'duration': parse_duration(video.get('duration')),
```
Comment (remitamine) [inline youtube_dl/extractor/rai.py:245]: is there a URL where the `duration` is available.

File: youtube_dl/extractor/rai.py [hunk 2]
```diff
@@ -262,29 +256,33 @@ class RaiPlayPlaylistIE(InfoExtractor):
...
+            playlist_id, 'Downloading program JSON')
+
+        title = media['name']
+        description = media['program_info']['description']
 
-        title = self._html_search_meta(
-            ('programma', 'nomeProgramma'), webpage, 'title')
-        description = unescapeHTML(self._html_search_meta(
-            ('description', 'og:description'), webpage, 'description'))
+        content_sets = [s['id'] for b in media['blocks'] for s in b['sets']]
```
Comment (remitamine) [inline youtube_dl/extractor/rai.py:274]: https://github.com/ytdl-org/youtube-dl#mandatory-and-optional-metafields

File: youtube_dl/extractor/rai.py [hunk 3]
```diff
@@ -262,29 +256,33 @@ class RaiPlayPlaylistIE(InfoExtractor):
...
-            video_url = urljoin(url, mobj.group('path'))
-            entries.append(self.url_result(
-                video_url, ie=RaiPlayIE.ie_key(),
-                video_id=RaiPlayIE._match_id(video_url)))
+        for cs in content_sets:
+            medias = self._download_json(
+                '%s/%s.json' % (urljoin(RaiBaseIE._BASE_URL, 'programmi/' + playlist_id), cs),
+                cs, 'Downloading content set JSON')
+            for m in medias['items']:
+                video_url = urljoin(url, m['path_id'])
```
Comment (remitamine) [inline youtube_dl/extractor/rai.py:282]: should not fail if the extraction of one set or item is not possible.

File: youtube_dl/extractor/rai.py [hunk 4]
```diff
@@ -240,20 +222,32 @@ class RaiPlayLiveIE(RaiBaseIE):
     def _real_extract(self, url):
         display_id = self._match_id(url)
 
-        webpage = self._download_webpage(url, display_id)
+        media = self._download_json(
+            '%s.json' % urljoin(RaiBaseIE._BASE_URL, 'dirette/' + display_id),
```
Comment (remitamine) [inline youtube_dl/extractor/rai.py:226]: capture the base URL in `_VALID_URL` regular expression and then add `.json` suffix(same for `RaiPlayPlaylistIE`).

**Review by remitamine — CHANGES_REQUESTED** [review]

File: youtube_dl/extractor/rai.py [hunk 5]
```diff
@@ -166,10 +148,11 @@ class RaiPlayIE(RaiBaseIE):
     }]
 
     def _real_extract(self, url):
-        url, video_id = re.match(self._VALID_URL, url).groups()
+        mobj = re.match(self._VALID_URL, url)
+        url, base, video_id, ext = mobj.group('url', 'base', 'id', 'ext')
 
         media = self._download_json(
-            url.replace('.html', '.json'), video_id, 'Downloading video JSON')
+            '%s%s.json' % (base, video_id), video_id, 'Downloading video JSON')
```
[the remark this answers is not in the document]
Comment (iamleot) [reply youtube_dl/extractor/rai.py:154]: It wouldn't catch the `.json` extension URLs.

File: youtube_dl/extractor/rai.py [hunk 6]
```diff
@@ -122,27 +123,8 @@ def _extract_subtitles(url, subtitle_url):
 
 
 class RaiPlayIE(RaiBaseIE):
-    _VALID_URL = r'(?P<url>https?://(?:www\.)?raiplay\.it/.+?-(?P<id>%s)\.html)' % RaiBaseIE._UUID_RE
+    _VALID_URL = r'(?P<url>(?P<base>https?://(?:www\.)?raiplay\.it/.+?-)(?P<id>%s)(?P<ext>\.(?:html|json)))' % RaiBaseIE._UUID_RE
```
[the remark this answers is not in the document]
Comment (iamleot) [reply youtube_dl/extractor/rai.py:125]: I will remove unused ones, thanks!

File: youtube_dl/extractor/rai.py (hunk 1 above)
Comment (iamleot) [reply youtube_dl/extractor/rai.py:245]: There should be... I will try to find one and add it as a test, thanks!

File: youtube_dl/extractor/rai.py (hunk 2 above)
Comment (iamleot) [reply youtube_dl/extractor/rai.py:274]: Thanks!  I will readd `title`!

File: youtube_dl/extractor/rai.py (hunk 1 above)
Comment (iamleot) [reply youtube_dl/extractor/rai.py:245]: Yes, the one in the test (<http://www.raiplay.it/video/2014/04/Report-del-07042014-cb27157f-9dd0-4aee-b788-b1f67643a391.html>) has it.

File: youtube_dl/extractor/rai.py (hunk 2 above)
Comment (iamleot) [reply youtube_dl/extractor/rai.py:274]: I have actually double-checked that and all mandatory metafields seems available... Can you please elaborate what's the problem with the quoted diff?

Comment (remitamine) [reply youtube_dl/extractor/rai.py:274]: > Any field apart from the aforementioned ones are considered optional. That means that extraction should be tolerant to situations when sources for these fields can potentially be unavailable (even if they are always available at the moment) and future-proof in order not to break the extraction of general purpose mandatory fields.

- the playlist title and description are not mandatory.
- extraction should not fail if any of the fields `blocks`, `sets`, and `id` are not available.

Comment (iamleot) [reply youtube_dl/extractor/rai.py:274]: Thanks for clarifying, I will adjust that!

File: youtube_dl/extractor/rai.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (iamleot) [reply youtube_dl/extractor/rai.py]: This should be addressed by https://github.com/ytdl-org/youtube-dl/pull/27077/commits/db01844f327b081e1921756aeae91deec0d5d747

File: youtube_dl/extractor/rai.py (hunk 2 above)
Comment (iamleot) [reply youtube_dl/extractor/rai.py]: This should be addressed by https://github.com/ytdl-org/youtube-dl/pull/27077/commits/1ac644a2a06e84c116e2c21d21a842251b8aeb25

File: youtube_dl/extractor/rai.py (hunk 3 above)
Comment (iamleot) [reply youtube_dl/extractor/rai.py]: This should be addressed by https://github.com/ytdl-org/youtube-dl/pull/27077/commits/12c3dae236c0ac5fdeb9ef97e07ac75699c61242

File: youtube_dl/extractor/rai.py (hunk 4 above)
Comment (iamleot) [reply youtube_dl/extractor/rai.py]: This should be addressed by https://github.com/ytdl-org/youtube-dl/pull/27077/commits/191d41a86a3ff846f132c7e2b97f4db6525bacb4

File: youtube_dl/extractor/rai.py (hunk 1 above)
Comment (remitamine) [reply youtube_dl/extractor/rai.py:245]: this URL you have provided is not related to this part of the code(`RaiPlayLiveIE`).

File: youtube_dl/extractor/rai.py [hunk 7]
```diff
@@ -238,53 +219,79 @@ class RaiPlayLiveIE(RaiBaseIE):
...
 
         entries = []
-        for mobj in re.finditer(
-                r'<a\b[^>]+\bhref=(["\'])(?P<path>/raiplay/video/.+?)\1',
-                webpage):
-            video_url = urljoin(url, mobj.group('path'))
-            entries.append(self.url_result(
-                video_url, ie=RaiPlayIE.ie_key(),
-                video_id=RaiPlayIE._match_id(video_url)))
+        for cs in content_sets:
```
Comment (remitamine) [inline youtube_dl/extractor/rai.py:284]: there is no need to collect the `content_sets` and then iterate over them, you can process them directly.

File: youtube_dl/extractor/rai.py [hunk 8]
```diff
@@ -238,53 +219,79 @@ class RaiPlayLiveIE(RaiBaseIE):
...
+            'duration': parse_duration(video.get('duration')),
         }
 
+        info.update(relinker_info)
+        return info
+
 
 class RaiPlayPlaylistIE(InfoExtractor):
-    _VALID_URL = r'https?://(?:www\.)?raiplay\.it/programmi/(?P<id>[^/?#&]+)'
+    _VALID_URL = r'(?P<base>https?://(?:www\.)?raiplay\.it/)programmi/(?P<id>[^/?#&]+)'
```
Comment (remitamine) [inline youtube_dl/extractor/rai.py:253]: the part you're matching is static and won't change, I meant to match to the end of the URL(the same for `RaiPlayLiveIE`).

**Review by remitamine — CHANGES_REQUESTED** [review]

File: youtube_dl/extractor/rai.py (hunk 1 above)
Comment (iamleot) [reply youtube_dl/extractor/rai.py]: Whoops, I accidentally confused them with RaiPlayIE, thanks!  I have just double-checked all such URL via <https://www.raiplay.it/dirette/> and none of them have an `alt_title` or `duration` so I removed both of them. This should be addressed by https://github.com/ytdl-org/youtube-dl/pull/27077/commits/2ea4cfb5f8d7a882c2e5c7aedd60af18a36881e5

File: youtube_dl/extractor/rai.py (hunk 7 above)
Comment (iamleot) [reply youtube_dl/extractor/rai.py]: Thanks!  I have refactored it and this should be fixed via https://github.com/ytdl-org/youtube-dl/pull/27077/commits/db472dd61bf49268727470be6ec40ee129d2bca1

File: youtube_dl/extractor/rai.py (hunk 8 above)
Comment (iamleot) [reply youtube_dl/extractor/rai.py:251]: Sorry but then I don't get it.  Please elaborate further.

Comment (iamleot) [reply youtube_dl/extractor/rai.py:251]: Do you meant to add possible trailing `/` or possible trailing extension (`.html` and/or `.json` if any)?

Comment (remitamine) [reply youtube_dl/extractor/rai.py:251]:
```suggestion
    _VALID_URL = r'(?P<base>https?://(?:www\.)?raiplay\.it/programmi/(?P<id>[^/?#&]+))'
```

Comment (iamleot) [reply youtube_dl/extractor/rai.py]: Thanks!

This should be addressed by https://github.com/ytdl-org/youtube-dl/pull/27077/commits/65600a1efd2891995a5fb6d96d32c77969b1a3ef

**Comment by iamleot** [comment]
If/when that looks good to be merged, please let me know that I will rebase and squash all commits in a single one and adjust commit message accordingly (there were something that I added initially but then amended based on review by @remitamine!).

Thanks!

**Comment by remitamine** [comment]
you can squash the commits and I will base the rest of the changes on this.

**Comment by iamleot** [comment]
Thanks!  I have just squashed all the commits to the only relevant one and updated the commit message.

If there is anything else to do or if possible testing (also in the future) is needed do not hesitate to contact me!

**Comment by iamleot** [comment]
remitamine writes:
> Merged #27077 into master.

Thanks!

**Pull request merged**